drivers/CpuUid.cpp

```diff
@@ -0,0 +1,81 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstdio>
+#include "hal/cpu_uid_api.h"
+#include "drivers/CpuUid.h"
+
+#if DEVICE_CPUUID
+
+namespace mbed {
+
+CpuUid::CpuUid() : _data(NULL)
+{
+    _size = cpu_uid_get_length();
+    if (0 < _size)
+    {
+        _data = new uint8_t[_size];
+        cpu_uid_get_uid(_data);
+    }
+}
+
+CpuUid::~CpuUid()
+{
+    if (_data)
+    {
+        delete _data;
+    }
+}
+
+CpuUid::operator std::string()
+{
+    std::string str;
+    char buf[3];
+    
+    for (int i = 0; i < _size; ++i)
+    {
+        snprintf(buf, 3, "%.2X", _data[i]);
```

Comment:
That is quite expensive in memory terms, in debug build it may not make much difference, but for release builds it may substanially inflate the binary.

Reply:
Do you mean the usage of std::string or the snprintf call?

As for (vsn)printf, this is already linked into all the classes which are derived from stream... for example the printf of the Serial class... I do not think another call (in the cpuid driver) will boost flash usage.... but please correct me if i'm misunderstanding something here ;)

Reply:
I just did some mememory meassurements with release profile on NUCLEO F411RE:
basic blinky: 46212B
using only uint8_t part of cpuuid: 46532B
using str operator: 62808B

That's in a situation when user doesn't use any form of printf elsewhere. In the end linker is pretty good in removing the symbols we don't use, but I still think we should have some kind of warning in the docs. Could you add a note warning user about possible substantial size growth due to pulling in sprintf and all the dependencies when using this operator.

Reply:
... or I could just write a small part of code which compensates the code size by the amount of my lazyness in the first commit ;)

stay tuned...